src/coreclr/src/vm/nativeimage.cpp

```diff
@@ -0,0 +1,125 @@
+
+#include <shlwapi.h>
+
+AssemblyNameIndex::AssemblyNameIndex(const SString& name, int32_t index)
+    : Name(name), Index(index)
+{
+    Name.Normalize();
+}
+
+NativeImage::NativeImage(
+    PEFile *pPeFile,
+    PEImage *pPeImage,
+    READYTORUN_HEADER *pHeader,
+    LPCUTF8 nativeImageName,
+    LoaderAllocator *pLoaderAllocator,
+    AllocMemTracker& amTracker)
+{
+    CONTRACTL
+    {
+        CONSTRUCTOR_CHECK;
+        STANDARD_VM_CHECK;
+        INJECT_FAULT(COMPlusThrowOM(););
+    }
+    CONTRACTL_END;
+
+    LoaderHeap *pHeap = pLoaderAllocator->GetHighFrequencyHeap();
+    m_pLoadContext = pPeFile->GetAssemblyLoadContext();
+    m_utf8SimpleName = nativeImageName;
+    m_pPeImage = pPeImage;
+    m_runEagerFixups = true;
+    
+    m_pReadyToRunInfo.Assign(new (amTracker.Track(pHeap->AllocMem((S_SIZE_T)sizeof(ReadyToRunInfo))))
+        ReadyToRunInfo(/*pModule*/ NULL, pPeImage->GetLoadedLayout(), pHeader, /*compositeImage*/ NULL, &amTracker), /*takeOwnership*/ TRUE);
+    m_pComponentAssemblies = m_pReadyToRunInfo->GetCompositeInfo()->FindSection(ReadyToRunSectionType::ComponentAssemblies);
+    m_componentAssemblyCount = m_pComponentAssemblies->Size / sizeof(READYTORUN_COMPONENT_ASSEMBLIES_ENTRY);
+    
+    // Check if the current module's image has native manifest metadata, otherwise the current->GetNativeAssemblyImport() asserts.
+    m_pManifestMetadata = pPeImage->GetNativeMDImport();
+
+    HENUMInternal assemblyEnum;
+    HRESULT hr = m_pManifestMetadata->EnumAllInit(mdtAssemblyRef, &assemblyEnum);
+    mdAssemblyRef assemblyRef;
+    int assemblyIndex = 0;
+    while (m_pManifestMetadata->EnumNext(&assemblyEnum, &assemblyRef))
+    {
+        LPCSTR assemblyName;
+        hr = m_pManifestMetadata->GetAssemblyRefProps(assemblyRef, NULL, NULL, &assemblyName, NULL, NULL, NULL, NULL);
+        m_assemblySimpleNameToIndexMap.Add(new (amTracker.Track(pHeap->AllocMem((S_SIZE_T)sizeof(AssemblyNameIndex))))
+            AssemblyNameIndex(SString(SString::Ansi, assemblyName), assemblyIndex));
+        assemblyIndex++;
+    }
+}
+
+bool NativeImage::Matches(LPCUTF8 utf8SimpleName, const AssemblyLoadContext *pLoadContext) const
+{
+    return pLoadContext == m_pLoadContext && !strcmp(utf8SimpleName, m_utf8SimpleName);
+}
+
+bool NativeImage::EagerFixupsNeedToRun()
+{
+    bool runEagerFixups = m_runEagerFixups;
+    m_runEagerFixups = false;
+    return runEagerFixups;
+}
+
+NativeImage *NativeImage::Open(
+    PEFile *pPeFile,
+    PEImage *pPeImage,
+    LPCUTF8 nativeImageName,
+    LoaderAllocator *pLoaderAllocator)
+{
+    READYTORUN_HEADER *pHeader = pPeImage->GetLoadedLayout()->GetReadyToRunHeader();
+    if (pHeader == NULL)
+    {
+        return NULL;
+    }
+    LoaderHeap *pHeap = pLoaderAllocator->GetHighFrequencyHeap();
+    AllocMemTracker amTracker;
+    NativeImage *result = new (amTracker.Track(pHeap->AllocMem((S_SIZE_T)sizeof(NativeImage))))
```

Comment:
We typically use loader allocator to only allocate types that only contain other items allocated by loader allocator. It is not the case here: NativeImage has multiple items that are allocated on regular heap. I think it would be better to allocated NativeImage on regular heap.

Reply:
To be precise, the rule is, never allocate via a LoaderHeap a C++ object that has an implicit or explicit destructor. NativeImage has an implicit destructor due to the presence of a NewHolder and an SHash.

---
In reply to: [384631579](https://github.com/dotnet/runtime/pull/32500#discussion_r384631579) [](ancestors = 384631579)

Reply:
Fixed in 11th commit.